File: 2022/day11/day11.py
class Queue:
    def __init__(self):
        self.queue = []
        self.size = 0

    def enqueue(self, el):
        self.queue.append(el)
        self.size += 1

    def dequeue(self):
        el = self.queue[0]
        self.queue = self.queue[1:]
        self.size -= 1
        return el

    def is_empty(self):
        return self.size == 0


class Monkey:
    def __init__(self, itms, opr, tst, tt, dv):
        self.items = itms
        self.op = opr
        self.test = tst
        self.to_throw = tt
        self.div = dv
        self.monkey_business = 0


def read_test_line(line):
    fields = line.split(' ')
    last = len(fields) - 1
    return int(fields[last])


def read_items(line):
    line = line.strip()
    fields = line.split(' ')[2:]

    q = Queue()
    for s in fields:
        off = len(s) - 1
        if s[off] != ',':
            off += 1

        n = int(s[:off])
        q.enqueue(n)

    return q


def read_operation_line(line):
    line = line.strip()
    fields = line.split(' ')[4:]

    n = None
    try:
        n = int(fields[1])
    except:
        n = -1

    if fields[0] == "+":
        if n < 0:
            return lambda x: x + x
        else:
            return lambda x: x + n
    elif fields[0] == "*":
        if n < 0:
            return lambda x: x * x
        else:
            return lambda x: x * n
    else:
        print('Operation not known')
        exit(2)


def parse_input(lines):
    monkeys = []
    for i in range(0, len(lines), 7):
        itms = read_items(lines[i+1])
        opr = read_operation_line(lines[i+2])
        div = read_test_line(lines[i+3])
        tst = eval(f'lambda x: x%{div} == 0')
        tt = {}
        tt[True] = read_test_line(lines[i+4])
        tt[False] = read_test_line(lines[i+5])
        
        monkeys.append(Monkey(itms, opr, tst, tt, div))
    return monkeys


def print_round(monkeys, round):
    print(f'After round {round}')
    for idx, m in enumerate(monkeys):
        print(f'Monkey {idx}: {m.items.queue}\tbusiness: {m.monkey_business}')
    print()


def part1(lines):
    monkeys = parse_input(lines)

    for _ in range(20):
        for idx, m in enumerate(monkeys):
            monkeys[idx].monkey_business += m.items.size
            while not m.items.is_empty():
                item = m.items.dequeue()
                item = m.op(item)
                item //= 3
                key = m.test(item)
                monkeys[m.to_throw[key]].items.enqueue(item)
        # print_round(monkeys, r)

    monkey_businesses = []
    for m in monkeys:
        monkey_businesses.append(m.monkey_business)
    monkey_businesses = sorted(monkey_businesses, reverse=True)

    print(f'The monkey business value is: {monkey_businesses[0] * monkey_businesses[1]}')


def part2(lines):
    monkeys = parse_input(lines)
    lcm = 1
    for m in monkeys:
        lcm *= m.div
    for r in range(10000):
        for idx, m in enumerate(monkeys):
            monkeys[idx].monkey_business += m.items.size
            while not m.items.is_empty():
                item = m.items.dequeue()
                item = m.op(item)
                key = m.test(item)
                monkeys[m.to_throw[key]].items.enqueue(item % lcm)
        if r % 100 == 0:
            print(f'Round {r}...')
        # print_round(monkeys, r)

    monkey_businesses = []
    for m in monkeys:
        monkey_businesses.append(m.monkey_business)
    monkey_businesses = sorted(monkey_businesses, reverse=True)

    print(f'The monkey business value is: {monkey_businesses[0] * monkey_businesses[1]}')


def main():
    lines = None
    with open('mini_input', 'r') as f:
        lines = f.read().splitlines()

    part1(lines)
    part2(lines)


if __name__ == '__main__':
    main()

File: 2022/day11/day11.go
package main

import (
  "io"
  "os"
  "fmt"
  "sort"
  "bufio"
  "strings"
  "strconv"
	
	queue "AdventOfCode/ds/queue"
)

const (
  FEW_ROUNDS = 20
  ROUNDS = 10000
  COOL_OFF = 3
)

type Monkey struct {
  items *queue.Queue[int]
  op func (int) int
  test func (int) bool
  toThrow map[bool]int
  monkeyBusiness int
  div int
}

func NewMonkey(itms *queue.Queue[int], opr func(int) int, tst func(int) bool, tt map[bool]int, div int) Monkey {
  return Monkey{itms, opr, tst, tt, 0, div}
}

func readTestLine(line string) int {
  fields := strings.Split(line, " ")
  last := len(fields) - 1
  n, err := strconv.Atoi(fields[last])
  if err != nil {
    panic("Cannot convert test result")
  }
  return n
}

func readItems(line string) *queue.Queue[int] {
  line = strings.TrimSpace(line)
  fields := strings.Split(line, " ")[2:]

  q := queue.NewQueue[int]()
  for _, nStr := range fields {
    off := len(nStr) - 1
    if nStr[off] != ',' {
      off++
    }

    n, err := strconv.Atoi(nStr[:off])
    if err != nil {
      fmt.Println("ERROR:", nStr[:off])
      panic("Cannot convert item")
    }
    q.Enqueue(n)
  }

  return q
}

func readOperationLine(line string) (func(int) int) {
  line = strings.TrimSpace(line)
  fields := strings.Split(line, " ")[4:]
  
  n, err := strconv.Atoi(fields[1])
  if err != nil {
    n = -1
  }

  switch fields[0] {
  case "+":
    if n < 0 {
      return func(x int) int {
        return x + x
      }
    }
    return func(x int) int {
      return x + n
    }
  case "*":
    if n < 0 {
      return func(x int) int {
        return x * x
      }
    }
    return func(x int) int {
      return x * n
    }
  default:
    panic("Operation not known")
  }
}

func parseInput(f *os.File) []Monkey {
  var lines []string
  f.Seek(0, io.SeekStart)
  scanner := bufio.NewScanner(f)
  for scanner.Scan() {
    lines = append(lines, scanner.Text())
  }

  n := len(lines)
  var monkeys []Monkey
  for i := 0; i < n; i += 7 {
    var itms *queue.Queue[int]
    tt := make(map[bool]int)

    itms = readItems(lines[i+1])
    opr := readOperationLine(lines[i+2])
    div := readTestLine(lines[i+3])
    tst := func (x int) bool {
      return x%div == 0
    }
    tt[true] = readTestLine(lines[i+4])
    tt[false] = readTestLine(lines[i+5])

    monkeys = append(monkeys, NewMonkey(itms, opr, tst, tt, div))
  }

  return monkeys
}

func printRound(monkeys []Monkey, round int) {
  fmt.Println("After round", round)
  for idx, m := range monkeys {
    fmt.Printf("Monkey (%d): %v\tbusiness: %d\n", idx, m.items.GetQueue(), m.monkeyBusiness)
  }
  fmt.Println()
}

func part1(monkeys []Monkey) {
  for i := 1; i <= FEW_ROUNDS; i++ {
    for idx, m := range monkeys {
      monkeys[idx].monkeyBusiness += m.items.GetSize()
      for !m.items.IsEmpty() {
        item := m.items.Dequeue()
        item = m.op(item)
        item /= COOL_OFF
        key := m.test(item)
        monkeys[m.toThrow[key]].items.Enqueue(item)
      }
    }
    // printRound(monkeys, i)
  }

  var monkeyBusinesses []int
  for _, m := range monkeys {
    monkeyBusinesses = append(monkeyBusinesses, m.monkeyBusiness)
  }
  sort.Ints(monkeyBusinesses)

  n := len(monkeyBusinesses)
  fmt.Println("The monkey business value is:", monkeyBusinesses[n-1] * monkeyBusinesses[n-2])
}

// // IT WORKS BUT TAKES TOO MUCH MEMORY AND TIME :( [using library math/big]
// func part2(monkeys []Monkey) {
//   for i := 1; i <= ROUNDS; i++ {
//     for idx, m := range monkeys {
//       monkeys[idx].monkeyBusiness += m.items.size
//       for !m.items.IsEmpty() {
//         item := m.items.Dequeue()
//         item = m.op(item)
//         key := m.test(item)
//         monkeys[m.toThrow[key]].items.Enqueue(item)
//       }
//     }
//     if i%100 == 0 {
//       fmt.Printf("Round %d...\n", i)
//     }
//     if i == 1 || i == 20 || i%1000 == 0 {
//       fmt.Printf("=== After round %d ===\n", i)
//       for idx, m := range monkeys {
//         fmt.Println("Monkey", idx, "inspected items", m.monkeyBusiness, "times")
//       }
//       fmt.Println()
//       printRound(monkeys, i)
//     }
//   }
//
//   var monkeyBusinesses []int
//   for _, m := range monkeys {
//     monkeyBusinesses = append(monkeyBusinesses, m.monkeyBusiness)
//   }
//   sort.Ints(monkeyBusinesses)
//
//   n := len(monkeyBusinesses)
//   fmt.Println("The monkey business value is:", monkeyBusinesses[n-1] * monkeyBusinesses[n-2])
// }

func part2(monkeys []Monkey) {
  // there is no need to carry a number greater than the LCM of the monkeys'
  // dividends, because the math modulo LCM is the same since we do not
  // divide anymore by COOL_OFF
  lcm := 1
  for _, m := range monkeys {
    lcm *= m.div
  }

  for i := 1; i <= ROUNDS; i++ {
    for idx, m := range monkeys {
      monkeys[idx].monkeyBusiness += m.items.GetSize()
      for !m.items.IsEmpty() {
        item := m.items.Dequeue()
        item = m.op(item)
        key := m.test(item)
        monkeys[m.toThrow[key]].items.Enqueue(item % lcm)
      }
    }
    // if i == 1 || i == 20 || i%1000 == 0 {
    //   fmt.Printf("=== After round %d ===\n", i)
    //   for idx, m := range monkeys {
    //     fmt.Println("Monkey", idx, "inspected items", m.monkeyBusiness, "times")
    //   }
    //   fmt.Println()
    //   printRound(monkeys, i)
    // }
  }
  
  var monkeyBusinesses []int
  for _, m := range monkeys {
    monkeyBusinesses = append(monkeyBusinesses, m.monkeyBusiness)
  }
  sort.Ints(monkeyBusinesses)

  n := len(monkeyBusinesses)
  fmt.Println("The monkey business value is:", monkeyBusinesses[n-1] * monkeyBusinesses[n-2])
}


func main() {
  f, err := os.Open("input")
  if err != nil {
    panic(err)
  }
  defer f.Close()

  part1(parseInput(f))
  part2(parseInput(f))
}

